File: api/src/main/java/com/ford/labs/retroquest/team/TeamService.java
/*
 * Copyright (c) 2022 Ford Motor Company
 * All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.ford.labs.retroquest.team;

import com.ford.labs.retroquest.actionitem.ActionItemRepository;
import com.ford.labs.retroquest.column.Column;
import com.ford.labs.retroquest.column.ColumnRepository;
import com.ford.labs.retroquest.exception.TeamDoesNotExistException;
import com.ford.labs.retroquest.thought.ThoughtRepository;
import com.ford.labs.retroquest.websocket.WebsocketService;
import org.springframework.dao.DataIntegrityViolationException;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

@Service
public class TeamService {
    private final ThoughtRepository thoughtRepository;
    private final ActionItemRepository actionItemRepository;
    private final TeamRepository teamRepository;
    private final PasswordEncoder passwordEncoder;
    private final ColumnRepository columnRepository;

    public TeamService(
        ThoughtRepository thoughtRepository,
        ActionItemRepository actionItemRepository,
        TeamRepository teamRepository,
        PasswordEncoder passwordEncoder,
        ColumnRepository columnRepository
    ) {
        this.thoughtRepository = thoughtRepository;
        this.actionItemRepository = actionItemRepository;
        this.teamRepository = teamRepository;
        this.passwordEncoder = passwordEncoder;
        this.columnRepository = columnRepository;
    }

    public Team getTeamByName(String teamName) {
        return teamRepository.findTeamByNameIgnoreCase(teamName.trim())
            .orElseThrow(TeamDoesNotExistException::new);
    }

    public Team getTeamByUri(String teamUri) {
        return teamRepository.findTeamByUri(teamUri.toLowerCase())
            .orElseThrow(TeamDoesNotExistException::new);
    }

    public String convertTeamNameToURI(String teamName) {
        return teamName.toLowerCase().replace(" ", "-");
    }

    public String encodePassword(String password) {
        return passwordEncoder.encode(password);
    }

    public Team createNewTeam(CreateTeamRequest createTeamRequest) {
        var encryptedPassword = this.encodePassword(createTeamRequest.getPassword());

        var trimmedName = createTeamRequest.getName().trim();
        var uri = convertTeamNameToURI(trimmedName);
        teamRepository
            .findTeamByUri(uri)
            .ifPresent(s -> {
                throw new DataIntegrityViolationException(s.getUri());
            });

        var team = new Team(
                uri,
                trimmedName,
                encryptedPassword
        );
        team = teamRepository.save(team);
        generateColumns(team);

        return team;
    }

    public void generateColumns(Team team) {
        var happyColumn = new Column(null, "happy", "Happy", team.getUri());
        var confusedColumn = new Column(null, "confused", "Confused", team.getUri());
        var unhappyColumn = new Column(null, "unhappy", "Sad", team.getUri());

        columnRepository.save(happyColumn);
        columnRepository.save(confusedColumn);
        columnRepository.save(unhappyColumn);
    }
}

File: ui/src/react/services/TeamService.ts
/*
 * Copyright (c) 2021 Ford Motor Company
 * All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import axios from 'axios';

import { AuthService } from '../../app/modules/auth/auth.service';

import { CREATE_TEAM_API_PATH, getTeamNameApiPath, LOGIN_API_PATH } from './ApiConstants';

function postThenSetTokenAndReturnTeamId(url: string, data: { name: string; password: string }) {
  return axios.post(url, data).then((response) => {
    const token = response.data;
    const teamId = response.headers.location;
    AuthService.setToken(token);
    return teamId;
  });
}

const TeamService = {
  login: (name: string, password: string): Promise<string> => {
    return postThenSetTokenAndReturnTeamId(LOGIN_API_PATH, { name, password });
  },

  create: (name: string, password: string): Promise<string> => {
    return postThenSetTokenAndReturnTeamId(CREATE_TEAM_API_PATH, { name, password });
  },

  getTeamName: (teamId: string): Promise<string> => {
    const TEAM_NAME_API_PATH = getTeamNameApiPath(teamId);
    return axios.get(TEAM_NAME_API_PATH).then((res) => res.data);
  },
};

export default TeamService;
